src/machine/vm.cr

```diff
@@ -1082,6 +1082,11 @@ module VM
         reg = Register.new stack_pop UInt8
         value = reg_read Int32, reg
         STDOUT.puts "#{value}"
+      when Syscall::READ
+        char = STDIN.raw &.read_char
+        if char.is_a?(Char)
+          reg_write Register::R0.byte, char.ord
```

Comment:
:+1:. I was using the standard Unix approach, where syscall arguments are in standard registers. I can change it, though I still want to understand better what is going on.

Let's consider the call to `puts` in in [examples/loops.asm](https://github.com/KCreate/stackvm/blob/b218f5408d3a0e3d92c2c3db6db39142f1b94913/examples/loops.asm#L31). It looks like `push t_register, counter` pushes an identifier corresponding to `r0` to the stack? Does this mean that the stack maintains types for its contents at runtime, to distinguish between numbers and register identifiers? This is very different from anything I've seen. Can you elaborate on the rationale for this design?

What does `push` do, decrement `sp` and write to the address in it?

I'm curious how my existing implementation is working. The following example doesn't provide a register identifier on the stack for `puts`, but it seems to work. `sys_puts` picks up the result of `sys_read` from `R0`.

```
.org 0x00
.label entry_addr
.label main

  push t_syscall, sys_read
  syscall
  push t_syscall, sys_puts
  syscall

  push t_syscall, sys_exit
  syscall
```